common/src/collision.rs

```diff
@@ -0,0 +1,477 @@
+        );
+        for cbb in self.bounding_boxes.iter() {
+            println!("\tA chunk");
+            //println!("\t\twith node id {}", cbb.node); // can't easily display node id
+            println!(
+                "\t\twith bounding box stretching from ({}, {}, {}) to ({}, {}, {})",
+                cbb.min_xyz[0],
+                cbb.min_xyz[1],
+                cbb.min_xyz[2],
+                cbb.max_xyz[0],
+                cbb.max_xyz[1],
+                cbb.max_xyz[2]
+            );
+        }
+    }
+}
+
+// translated_position should be the object position in the node coordinates of the chunk.
+// node can be easily factored out if it stops being convienent.
+impl ChunkBoundingBox {
+    pub fn get_chunk_bounding_box(
+        node: NodeId,
+        chunk: Vertex,
+        translated_position: na::Vector4<f64>,
+        radius: f64,
+        dimension: u8,
+    ) -> Option<Self> {
+        let euclidean_position =
+            (chunk.chunk_to_node().try_inverse().unwrap() * translated_position).xyz();
+        let mut min_xyz = na::Vector3::<u32>::new(0_u32, 0_u32, 0_u32);
+        let mut max_xyz = na::Vector3::<u32>::new(0_u32, 0_u32, 0_u32);
+
+        // It's important to note that euclidean_position is measured as chunk lengths, and radius is measured in absolute units.
+        // By coicidence, an absolute unit is aproximately a chunk's diameter, and only because of that there is no unit conversion here.
+
+        // verify at least one box corner is within the chunk
+        if euclidean_position
+            .iter()
+            .all(|n| n + radius > 0_f64 && n - radius < 1_f64)
+        {
+            min_xyz.x =
+                ((euclidean_position.x - radius).max(0_f64) * dimension as f64).floor() as u32;
+            max_xyz.x =
+                ((euclidean_position.x + radius).min(1_f64) * dimension as f64).ceil() as u32;
+
+            min_xyz.y =
+                ((euclidean_position.y - radius).max(0_f64) * dimension as f64).floor() as u32;
+            max_xyz.y =
+                ((euclidean_position.y + radius).min(1_f64) * dimension as f64).ceil() as u32;
+
+            min_xyz.z =
+                ((euclidean_position.z - radius).max(0_f64) * dimension as f64).floor() as u32;
+            max_xyz.z =
+                ((euclidean_position.z + radius).min(1_f64) * dimension as f64).ceil() as u32;
+            Some(ChunkBoundingBox {
+                node,
+                chunk,
+                min_xyz,
+                max_xyz,
+                dimension,
+            })
+        } else {
+            None
+        }
+    }
+
+    pub fn every_voxel<'b>(&'b self) -> impl Iterator<Item = u32> + 'b {
+        (self.min_xyz[2]..self.max_xyz[2]).flat_map(move |z| {
+            (self.min_xyz[1]..self.max_xyz[1]).flat_map(move |y| {
+                (self.min_xyz[0]..self.max_xyz[0]).map(move |x| {
+                    z + (self.dimension as u32) * y + (self.dimension as u32).pow(2) * x
```

Comment:
The data structure storing voxels has a 1-block margin, so it acts like an array where each dimension is `self.dimension + 2`. I believe this is one of the two main sources of error for the collision checking.
```suggestion
                    (x + 1) + lwm * (y + 1) + lwm.pow(2) * (z + 1) // Where lwm is (self.dimension as u32 + 2)
```